### PR TITLE
Propagate redirect_uri during device trust web auth

### LINKED ISSUE #46980: Device Trust interstitial "swallows" app deep link from URL
Expected behavior:

When you click through to an application subdomain behind Teleport App access with a URL (like https://app.teleport.example.com/path/to/something) it should preserve this URL and redirects the user.

Current behavior:

This works if Device Trust is not enabled.

When Device Trust is enabled, the interstitial dialog that prompts you to launch Teleport Connect for an authenticated session "swallows" the URL so the redirection does not happen - you get dumped on the root page of the URL (i.e. https://app.teleport.example.com)

This is a frustrating experience.

Bug details:
- Teleport version: 16.4.0

### PR DESCRIPTION
**Pull request by avatus**
In the web UI, if a user tries to access a page but they are unauthenticated, after logging in, they are automatically redirected back to the originally requested page. If the user has to authenticate their session with device trust, this URI got swallowed and the user would always end up at "/web". This PR passes the redirect through to the authentication process.

Fixes: https://github.com/gravitational/teleport/issues/46980

### CODE REVIEW

File: lib/web/device_trust.go [hunk 1]
```diff
@@ -76,7 +81,29 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
 
 	// Always redirect back to the dashboard, regardless of outcome.
 	app.SetRedirectPageHeaders(w.Header(), "" /* nonce */)
-	http.Redirect(w, r, "/web", http.StatusSeeOther)
+
+	redirectTo := "/web"
+
+	if redirectParam != "" {
+		parsedURL, err := url.Parse(redirectParam)
```
Comment (codingllama) [inline lib/web/device_trust.go:88]: Log non-nil errors?

File: lib/web/device_trust.go [hunk 2]
```diff
@@ -44,6 +47,8 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
 	confirmToken := &devicepb.DeviceConfirmationToken{}
 	confirmToken.Id = query.Get("id")
 	confirmToken.Token = query.Get("token")
+	redirectParam := query.Get("redirect_uri")
```
Comment (codingllama) [inline lib/web/device_trust.go:50]: nit:

```suggestion
	redirectURI := query.Get("redirect_uri")
```

so it matches the param?

File: lib/web/device_trust.go [hunk 3]
```diff
@@ -76,7 +81,29 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
...
 	// Always redirect back to the dashboard, regardless of outcome.
 	app.SetRedirectPageHeaders(w.Header(), "" /* nonce */)
-	http.Redirect(w, r, "/web", http.StatusSeeOther)
+
+	redirectTo := "/web"
+
+	if redirectParam != "" {
+		parsedURL, err := url.Parse(redirectParam)
+		if err == nil {
+			cleanPath := path.Clean(parsedURL.Path)
```
Comment (codingllama) [inline lib/web/device_trust.go:90]: This block seems like it could be a small helper function of its own.

File: lib/web/device_trust.go [hunk 4]
```diff
@@ -76,7 +81,29 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
...
+			// helps in situations where there is no path such as https://example.com
+			if cleanPath == "." || cleanPath == "" {
```
Comment (codingllama) [inline lib/web/device_trust.go:92]: Do we need to handle ".." too?

Comment (codingllama) [inline lib/web/device_trust.go:92]: path.Clean doesn't return an empty string:

> If the result of this process is an empty string, Clean returns the string ".".

https://pkg.go.dev/path#Clean

File: lib/web/device_trust.go [hunk 5]
```diff
@@ -76,7 +81,29 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
...
+				cleanPath = "/"
+			} else if !strings.HasPrefix(cleanPath, "/") {
+				cleanPath = "/" + cleanPath
+			}
+			// Prepend "/web" only if it's not already present
+			if !strings.HasPrefix(cleanPath, "/web") {
+				redirectTo = path.Join("/web", cleanPath)
+			} else {
+				redirectTo = cleanPath
+			}
```
Comment (codingllama) [inline lib/web/device_trust.go:102]: Let's make sure we test all the corner cases of this block. It seems a bit tricky to get right.

File: lib/web/device_trust_test.go [hunk 6]
```diff
@@ -36,85 +36,105 @@ import (
...
+			expectedRedirectTo: "/web",
+		},
+		{
+			name:               "with external redirect_uri",
+			redirectURI:        "https://example.com/path",
+			expectedRedirectTo: "/web/path",
+		},
+		{
+			name:               "with empty path redirect_uri",
+			redirectURI:        "https://example.com",
```
Comment (codingllama) [inline lib/web/device_trust_test.go:66]: What happens when the initial URL is `https://myapp.example.com` ?

File: lib/web/device_trust_test.go [hunk 7]
```diff
@@ -36,85 +36,105 @@ import (
...
+			fakeDevices := &fakeDevicesClient{}
+			wPack := newWebPack(
+				t,
+				1, /* numProxies */
+				withDevicesClientOverride(fakeDevices),
+			)
+			proxy := wPack.proxies[0]
+			aPack := proxy.authPack(t, "llama", nil /* roles */)
+			webClient := aPack.clt
+			ctx := context.Background()
```
Comment (codingllama) [inline lib/web/device_trust_test.go:84]: Could we use a single "pack" for all tests?

File: lib/web/device_trust_test.go [hunk 8]
```diff
@@ -36,85 +36,105 @@ import (
...
-					Token: "my-token-token",
+
+			req, err := http.NewRequestWithContext(ctx, "GET", webClient.Endpoint("webapi", "devices", "webconfirm"), nil)
+			require.NoError(t, err, "NewRequestWithContext failed")
+			req.URL.RawQuery = query.Encode()
+
+			resp, err := httpClient.Do(req)
+			require.NoError(t, err, "GET /webapi/devices/webconfirm failed")
+			defer resp.Body.Close()
+			io.Copy(io.Discard, resp.Body)
```
Comment (codingllama) [inline lib/web/device_trust_test.go:111]: We can straight up close the body here, it looks like.

```suggestion
			io.Copy(io.Discard, resp.Body)
			resp.Body.Close()
```

**Review by codingllama — COMMENTED** [review]
Thanks, Michael!

Could you make sure app redirects work as intended, and that we have test cases that mimic the scenario?

File: lib/web/device_trust.go (hunk 4 above)
Comment (avatus) [reply lib/web/device_trust.go:92]: `Clean` will only return the path, "/" or ".", with all ".." removed so I don't believe we need the check here. I can also remove the empty string check, thank you!

File: lib/web/device_trust_test.go (hunk 6 above)
Comment (avatus) [reply lib/web/device_trust_test.go:66]: When a user requests an app url and they aren't authenticated, they get redirected with a redirect uri of `https://example.com/web/launch/myapp.example.com` so this is caught the same as the other URLs. but I will add a test for that anyways. Thank you!

File: lib/web/device_trust.go (hunk 3 above)
Comment (avatus) [reply lib/web/device_trust.go:90]: Added a helper and tests

File: lib/web/device_trust.go (hunk 4 above)
Comment (codingllama) [reply lib/web/device_trust.go:92]: path.Clean("..") does return "..", so there is that to consider.

File: lib/web/device_trust_test.go (hunk 7 above)
Comment (avatus) [reply lib/web/device_trust_test.go]: I'll inspect more on Monday, but moving the packs outside of the tests caused them to fail. I'm its something simple I'm not seeing.

File: lib/web/device_trust.go [hunk 9]
```diff
@@ -76,7 +81,39 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
 
 	// Always redirect back to the dashboard, regardless of outcome.
 	app.SetRedirectPageHeaders(w.Header(), "" /* nonce */)
-	http.Redirect(w, r, "/web", http.StatusSeeOther)
+
+	redirectTo, err := h.getRedirectURL(redirectURI)
+	if err != nil {
+		h.log.Debugf("unable to parse redirectURI %s", redirectURI)
```
Comment (codingllama) [inline lib/web/device_trust.go:87]: Use structured logging:

```suggestion
		h.log.
			WithError(err).
			WithField("redirect_uri", redirectURI).
			Debug("Unable to parse redirectURI")
```

File: lib/web/device_trust.go [hunk 10]
```diff
@@ -76,7 +81,39 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
...
+	}
+	http.Redirect(w, r, redirectTo, http.StatusSeeOther)
 
 	return nil, nil
 }
+
+// getRedirectURL tries to parse the given redirectURI. It will always return a redirect url
+// even if the parse fails (in case of failture, the returned string is "/web")
+func (h *Handler) getRedirectURL(redirectURI string) (string, error) {
+	if redirectURI == "" {
```
Comment (codingllama) [inline lib/web/device_trust.go:97]: Suggestion:

```suggestion
	const basePath = "/web"
	if redirectURI == "" {
```

and then use it throughout?

File: lib/web/device_trust_test.go [hunk 11]
```diff
@@ -164,3 +170,76 @@ func (f *fakeDevicesClient) resetConfirmRequests() []*devicepb.ConfirmDeviceWebA
...
+			name:           "with completely invalid URI",
+			redirectURI:    "not a URI at all",
+			expectedURL:    "/web",
+			expectedErrMsg: "parse \"not a URI at all\": invalid URI for request",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result, err := h.getRedirectURL(tt.redirectURI)
```
Comment (codingllama) [inline lib/web/device_trust_test.go:232]: Isn't this tested as part of TestHandler_DeviceWebConfirm? If we can test via the public method that typically makes for a better test than testing a private helper directly.

File: lib/web/device_trust_test.go [hunk 12]
```diff
@@ -221,10 +221,9 @@ func TestHandler_GetRedirectURL(t *testing.T) {
 			expectedURL: "/web/existing/path",
 		},
 		{
-			name:           "with completely invalid URI",
-			redirectURI:    "not a URI at all",
-			expectedURL:    "/web",
-			expectedErrMsg: "parse \"not a URI at all\": invalid URI for request",
+			name:        "with completely invalid URI",
+			redirectURI: "not a URI at all",
+			expectedURL: "/web/not a URI at all",
```
Comment (codingllama) [inline lib/web/device_trust_test.go:226]: Shouldn't this go to /web?

File: lib/web/device_trust_test.go (hunk 11 above)
Comment (avatus) [reply lib/web/device_trust_test.go]: Removed this test and added missing cases to `TestHandler_DeviceWebConfirm`

**Review by codingllama — APPROVED** [review]

File: web/packages/shared/components/AuthorizeDeviceWeb/AuthorizeDeviceWeb.tsx [hunk 13]
```diff
@@ -36,10 +36,13 @@ import { makeDeepLinkWithSafeInput } from 'shared/deepLinks';
...
   const ctx = useTeleport();
+  const { search } = useLocation();
   const { id, token } = useParams<{
     id: string;
     token: string;
   }>();
+  const redirect = new URLSearchParams(search).get('redirect_uri');
+
   const { cluster, username } = ctx.storeUser.state;
   const deviceTrustAuthorize = makeDeepLinkWithSafeInput({
```
Comment (ravicious) [inline web/packages/shared/components/AuthorizeDeviceWeb/AuthorizeDeviceWeb.tsx:47]: The comment for this function says:

> `makeDeepLinkWithSafeInput` creates a deep link by interpolating passed arguments.
> Important: This function does not perform any validation or parsing. It must not be called with user-generated content.

This comment is no longer valid. IIRC my original intention, back when we didn't have search params here, was that `proxyHost`, `path` and `username` were simply concatenated together. So if we passed user-generated content there, it'd be possible to manipulate the function to generate some weird URLs.

With search params this is not a concern, since each param and its value get individually encoded. Could you update the comment so that it's up to date?

File: web/packages/shared/deepLinks.ts [hunk 14]
```diff
@@ -67,7 +67,7 @@ export type ConnectMyComputerDeepURL = BaseDeepURL & {
 
 export type AuthenticateWebDeviceDeepURL = BaseDeepURL & {
   pathname: '/authenticate_web_device';
-  searchParams: { id: string; token: string };
+  searchParams: { id: string; token: string; redirect?: string };
```
Comment (ravicious) [inline web/packages/shared/deepLinks.ts:70]:
```suggestion
  searchParams: { id: string; token: string; redirect_uri?: string };
```

Thoughts on renaming this param? I'd rather have it consistent instead of calling it `redirect_uri` in the Web UI but `redirect` in the deep link.

File: lib/web/device_trust.go [hunk 15]
```diff
@@ -44,6 +47,8 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
 	confirmToken := &devicepb.DeviceConfirmationToken{}
 	confirmToken.Id = query.Get("id")
 	confirmToken.Token = query.Get("token")
+	redirectURI := query.Get("redirect_uri")
```
Comment (ravicious) [inline lib/web/device_trust.go:50]: I think it's way to easy to mix up `redirectURI` (which could potentially lead to an external site) with `redirectTo` (which is safe).

Thoughts on renaming it to something like this?

```suggestion
	unsafeRedirectURI := query.Get("redirect_uri")
```

File: lib/web/device_trust.go [hunk 16]
```diff
@@ -76,7 +81,44 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
...
+			Debug("Unable to parse redirectURI")
+	}
+	http.Redirect(w, r, redirectTo, http.StatusSeeOther)
 
 	return nil, nil
 }
+
+// getRedirectURL tries to parse the given redirectURI. It will always return a redirect url
+// even if the parse fails (in case of failture, the returned string is "/web")
+func (h *Handler) getRedirectURL(redirectURI string) (string, error) {
```
Comment (ravicious) [inline lib/web/device_trust.go:99]:
```suggestion
func (h *Handler) getRedirectPath(redirectURI string) (string, error) {
```

Since I guess technically this returns just the path?

File: lib/web/device_trust.go [hunk 17]
```diff
@@ -76,7 +81,44 @@ func (h *Handler) deviceWebConfirm(w http.ResponseWriter, r *http.Request, _ htt
...
+	parsedURL, err := url.Parse(redirectURI)
+	if err != nil {
+		return basePath, trace.Wrap(err)
+	}
+
+	cleanPath := path.Clean(parsedURL.Path)
+	// helps in situations where there is no path such as https://example.com
+	if cleanPath == "." || cleanPath == ".." {
+		cleanPath = "/"
+	} else if !strings.HasPrefix(cleanPath, "/") {
```
Comment (ravicious) [inline lib/web/device_trust.go:115]: In what situation is `cleanPath` not equal to `.` or `..` and  also doesn't have the `/` prefix?

https://go.dev/play/p/bOEv7v_qAb5

File: lib/web/apiserver.go [hunk 18]
```diff
@@ -2050,9 +2050,11 @@ func (h *Handler) githubCallback(w http.ResponseWriter, r *http.Request, p httpr
...
 			// if a device web token is present, we must send the user to the device authorize page
 			// to upgrade the session.
-			// TODO (avatus) the web client currently doesn't handle any redirects after authorizing a web
-			// session with device trust. Once it does, append a redirect_url here as a query parameter
-			return fmt.Sprintf("/web/device/authorize/%s/%s", dwt.Id, dwt.Token)
+			redirectPath, err := BuildDeviceWebRedirectPath(dwt, res.ClientRedirectURL)
+			if err == nil {
+				return redirectPath
+			}
+			logger.WithError(err).Debug("Could not build redirect path")
```
Comment (ravicious) [inline lib/web/apiserver.go:2057]: I feel like this should be an error-level message, no? This has potential to fail is outside of our control (since the response comes from GitHub), so I suppose if something goes wrong, the admin cluster would like to know that without having to turn on debug logs.

```suggestion
			logger.WithError(err).Error("Could not build redirect path")
```

Comment (ravicious) [inline lib/web/apiserver.go:2057]: Is the intention to redirect to `res.ClientRedirectURL` anyway if `BuildDeviceWebRedirectPath` fails? I wonder if it'd help to be a little more explicit and do a normal `if err != nil` instead of depending on `return res.ClientRedirectURL` below – if the code gets moved or otherwise refactored, it'll be easier to keep the correct behavior.

File: web/packages/teleport/src/Login/useLogin.ts [hunk 19]
```diff
@@ -134,7 +134,12 @@ type LoginResponse = {
 };
 
 function authorizeWithDeviceTrust(token: DeviceWebToken) {
-  const authorize = cfg.getDeviceTrustAuthorizeRoute(token.id, token.token);
+  let redirect = history.getRedirectParam();
```
Comment (ravicious) [inline web/packages/teleport/src/Login/useLogin.ts:137]: I don't like the fact that  we pass around the whole URL, even though we end up using only the path, but I guess it's a tech debt from how the Web UI does app redirects, right?

**Review by ravicious — COMMENTED** [review]
I'll try to give it a go tomorrow, I managed to just look through the code today.

File: web/packages/shared/components/AuthorizeDeviceWeb/AuthorizeDeviceWeb.tsx
[the remark this answers is not in the document]
Comment (avatus) [reply web/packages/shared/components/AuthorizeDeviceWeb/AuthorizeDeviceWeb.tsx:1]: I think we were planning on maybe using some version of it in OSS as well and that hasn't worked out yet. But i'll just leave it here to avoid making more `e` changes + `e` ref updates if changes are needed in the future

File: web/packages/teleterm/src/deepLinks.ts [hunk 20]
```diff
@@ -124,6 +125,10 @@ export function parseDeepLink(rawUrl: string): DeepLinkParseResult {
           token,
         },
       };
+      // passing it directly above results in the string 'null' as the redirect param instead of ignoring it
```
[the remark this answers is not in the document]
Comment (avatus) [reply web/packages/teleterm/src/deepLinks.ts:128]: sure thats fair. i updated the test and filtered out null values when creating `searchParamString` in `makeDeepLinkSafeWithInput`

File: web/packages/teleport/src/config.ts [hunk 21]
```diff
@@ -475,8 +475,17 @@ const cfg = {
...
-    return generatePath(cfg.routes.deviceTrustAuthorize, { id, token });
+  getDeviceTrustAuthorizeRoute(id: string, token: string, redirect?: string) {
+    let path = generatePath(cfg.routes.deviceTrustAuthorize, {
+      id,
+      token,
+    });
+
+    if (redirect) {
+      path = `${path}?redirect_uri=${redirect}`;
+    }
```
[the remark this answers is not in the document]
Comment (avatus) [reply web/packages/teleport/src/config.ts:486]: agreed its inconsistent, its just a pattern thats done a few times here. ill update this specific method to use `URLSearchParams` instead

File: lib/web/device_trust.go (hunk 17 above)
Comment (avatus) [reply lib/web/device_trust.go:115]: SSO redirect may return something like `example/path` (I haven't seen it in actual testing but the fact that that param _could_ exist caused me to add this check).

The client redirect URLs from SSO callbacks generally _don't_ have the hostname attached and are only the paths, so I tried to catch it all (and added tests for them)

File: lib/web/apiserver.go (hunk 18 above)
Comment (codingllama) [reply lib/web/apiserver.go:2057]: If we are bumping the level for the cluster admin, then let's try to phrase the message in a way that they would know what to do here. The error is quite mysterious otherwise.

File: web/packages/teleport/src/Login/useLogin.ts (hunk 19 above)
Comment (avatus) [reply web/packages/teleport/src/Login/useLogin.ts:137]: Yeah this is just how its passed around unfortunately. I can add a TODO to fix later and inspect how many spots it will touch

File: lib/web/apiserver.go (hunk 18 above)
Comment (avatus) [reply lib/web/apiserver.go:2057]: The only error even returned from here is just "bad param" of not passing in a valid device trust web token and nothing to do with the redirect path actually. This seems like it will be a red herring regardless.

Comment (avatus) [reply lib/web/apiserver.go:2057]: I think it might be better to return a default "/web" route here if the `dtw` is invalid, log that its invalid, and then continue on our way imo. that way we are always returning from this method instead of dropping out due to an error. thoughts?

Comment (codingllama) [reply lib/web/apiserver.go:2057]: SGTM.

File: web/packages/shared/components/AuthorizeDeviceWeb/AuthorizeDeviceWeb.tsx (hunk 13 above)
Comment (avatus) [reply web/packages/shared/components/AuthorizeDeviceWeb/AuthorizeDeviceWeb.tsx:47]: I'm unsure your expected outcome for this comment to make it up to date besides calling "this must not be called with user-generated content". I removed that since I'm passing in user generated content now but if I'm missing the mark here, let me know.

File: web/packages/shared/utils/processRedirectUri.ts [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+ * @example
+ * processRedirectURI('/web', null) // returns '/web'
+ * processRedirectURI('/web', 'https://example.com/path') // returns '/web/path'
+ * processRedirectURI('/web', '/custom/path') // returns '/web/custom/path'
+ * processRedirectURI('/web', '/web/existing/path') // returns '/web/existing/path'
+ * processRedirectURI('/web', 'invalid://url') // returns '/web'
+ * processRedirectURI('/app', 'https://example.com/path') // returns '/app/path'
+ */
+export function processRedirectURI(
+  basePath: string,
```
Comment (avatus) [inline web/packages/shared/utils/processRedirectUri.ts:42]: the purpose of adding in the `basePath` here was so I could pass in something like `cfg.routes.root` and if that ever changed, this would be updated with it. However, I wasn't able to access that `cfg` everywhere this is used and had to manually pass `"/web"` anyway so... I could be convinced to just remove it and hard code in the `"/web"` . I really dont see that changing anytime anyway.

File: web/packages/shared/deepLinks.ts [hunk 23]
```diff
@@ -106,6 +105,8 @@ export function makeDeepLinkWithSafeInput<
     : '';
 
   const searchParamsString = Object.entries(args.searchParams)
+    // filter out params that have no value to prevent a string like "&myparam=null"
+    .filter(kv => !!kv[1])
```
Comment (gzdunek) [inline web/packages/shared/deepLinks.ts:109]: That will filter out `0`, `false` etc as well.
I'd check for `null` and `undefined` explicitly here.

File: web/packages/shared/utils/processRedirectUri.ts [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+    let path = url.pathname;
+    // If it already starts with basePath, return as is
+    if (path.startsWith(basePath)) {
+      return path;
+    }
+    if (path === '/') {
+      return basePath;
+    }
+    // For all other cases, ensure it starts with basePath
+    return path.startsWith(basePath)
```
Comment (gzdunek) [inline web/packages/shared/utils/processRedirectUri.ts:59]: Don't we already handle this case in the line 52.?

File: web/packages/shared/utils/processRedirectUri.ts [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+export function processRedirectURI(
+  basePath: string,
+  redirectURI: string | null
+): string {
+  if (!redirectURI) {
+    return basePath;
+  }
+  try {
+    const url = new URL(redirectURI);
+    let path = url.pathname;
```
Comment (gzdunek) [inline web/packages/shared/utils/processRedirectUri.ts:50]:
```suggestion
    const path = url.pathname;
```

File: web/packages/shared/utils/processRedirectUri.ts [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * @example
+ * processRedirectURI('/web', null) // returns '/web'
+ * processRedirectURI('/web', 'https://example.com/path') // returns '/web/path'
+ * processRedirectURI('/web', '/custom/path') // returns '/web/custom/path'
+ * processRedirectURI('/web', '/web/existing/path') // returns '/web/existing/path'
+ * processRedirectURI('/web', 'invalid://url') // returns '/web'
+ * processRedirectURI('/app', 'https://example.com/path') // returns '/app/path'
+ */
+export function processRedirectURI(
```
Comment (gzdunek) [inline web/packages/shared/utils/processRedirectUri.ts:41]: Nit: I'd stick to our JS conventions and call it `processRedirectUri` (same for the argument).

File: web/packages/shared/utils/processRedirectUri.ts (hunk 22 above)
Comment (gzdunek) [reply web/packages/shared/utils/processRedirectUri.ts:42]: I guess we can hardcode it? I'd only leave a comment that the value is equal to `cfg.routes.root`.

**Review by gzdunek — COMMENTED** [review]
I left a few comments, but I feel I need a fresh head for that :)
I will give it another look in the morning.

File: web/packages/shared/utils/processRedirectUri.ts [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/**
```
Comment (gzdunek) [inline web/packages/shared/utils/processRedirectUri.ts:18]: Let's add a new line here.

File: web/packages/shared/utils/processRedirectUri.ts [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+ * processRedirectURI('/web', null) // returns '/web'
+ * processRedirectURI('/web', 'https://example.com/path') // returns '/web/path'
+ * processRedirectURI('/web', '/custom/path') // returns '/web/custom/path'
+ * processRedirectURI('/web', '/web/existing/path') // returns '/web/existing/path'
+ * processRedirectURI('/web', 'invalid://url') // returns '/web'
+ * processRedirectURI('/app', 'https://example.com/path') // returns '/app/path'
+ */
+export function processRedirectUri(redirectUri: string | null): string {
+  // should be equal to cfg.routes.root
+  const basePath = '/web';
```
Comment (gzdunek) [inline web/packages/shared/utils/processRedirectUri.ts:43]: We can move it out of the function and call `BASE_PATH`.

File: web/packages/shared/utils/processRedirectUri.ts [hunk 29]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * This function handles various cases:
+ * - Null or empty input: Returns the basePath
+ * - Full URLs:
+ *   - External: Uses only the pathname, prepending the basePath if not already present
+ *   - Internal: Prepends the basePath if not already present
+ * - Relative paths: Prepends the basePath if not already present
+ * - Invalid URIs: Returns the basePath
+ *
+ * @param basePath - The base path to prepend to the processed URI (e.g., '/web')
```
Comment (gzdunek) [inline web/packages/shared/utils/processRedirectUri.ts:29]: We no longer have this param.

File: web/packages/shared/utils/processRedirectUri.ts
Comment (gzdunek) [inline web/packages/shared/utils/processRedirectUri.ts:1]: Having a `utils` directory is not the best pattern, as it tends to become a place for keeping unrelated code.
What do you think of moving this file and the test to `shared/redirects` directory? 
I'd reexport the function from `index.ts` so the callisites would have a nice import:
```ts
import { processRedirectUri } from 'shared/redirects';
```

**Review by gzdunek — APPROVED** [review]
